api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,723 @@
+	"m3.2xlarge":  30 * bytesize.GiB,
+	"c4.large":    3.75 * bytesize.GiB,
+	"c4.xlarge":   7.5 * bytesize.GiB,
+	"c4.2xlarge":  15 * bytesize.GiB,
+	"c4.4xlarge":  30 * bytesize.GiB,
+	"c4.8xlarge":  60 * bytesize.GiB,
+	"c3.large":    3.75 * bytesize.GiB,
+	"c3.xlarge":   7.5 * bytesize.GiB,
+	"c3.2xlarge":  15 * bytesize.GiB,
+	"c3.4xlarge":  30 * bytesize.GiB,
+	"c3.8xlarge":  60 * bytesize.GiB,
+	"g2.2xlarge":  15 * bytesize.GiB,
+	"g2.8xlarge":  60 * bytesize.GiB,
+	"x1.32xlarge": 1952 * bytesize.GiB,
+	"r3.large":    15.25 * bytesize.GiB,
+	"r3.xlarge":   30.5 * bytesize.GiB,
+	"r3.2xlarge":  61 * bytesize.GiB,
+	"r3.4xlarge":  122 * bytesize.GiB,
+	"r3.8xlarge":  244 * bytesize.GiB,
+	"i3.large":    15.25 * bytesize.GiB,
+	"i3.xlarge":   30.5 * bytesize.GiB,
+	"i3.2xlarge":  61 * bytesize.GiB,
+	"i3.4xlarge":  122 * bytesize.GiB,
+	"i3.8xlarge":  244 * bytesize.GiB,
+	"d2.xlarge":   30.5 * bytesize.GiB,
+	"d2.2xlarge":  61 * bytesize.GiB,
+	"d2.4xlarge":  122 * bytesize.GiB,
+	"d2.8xlarge":  244 * bytesize.GiB,
+}
 
 type EnvironmentScaler struct {
+	Client              *awsc.Client
+	EnvironmentProvider provider.EnvironmentProvider
+	ServiceProvider     provider.ServiceProvider
+	TaskProvider        provider.TaskProvider
+	JobStore            job.Store
+	Config              config.APIConfig
+	deployCache         map[string][]scaler.ResourceConsumer
 }
 
-func NewEnvironmentScaler() *EnvironmentScaler {
-	return &EnvironmentScaler{}
+func NewEnvironmentScaler(a *awsc.Client, e provider.EnvironmentProvider, s provider.ServiceProvider, t provider.TaskProvider, j job.Store, c config.APIConfig) *EnvironmentScaler {
+	return &EnvironmentScaler{
+		Client:              a,
+		EnvironmentProvider: e,
+		ServiceProvider:     s,
+		TaskProvider:        t,
+		JobStore:            j,
+		Config:              c,
+		// deployCache: map[string][]*Resource,
+	}
 }
 
 func (e *EnvironmentScaler) Scale(environmentID string) error {
-	return fmt.Errorf("EnvironmentScaler not implemented")
+	clusterName := addLayer0Prefix(e.Config.Instance(), environmentID)
+
+	resourceConsumers, resourceProviders, err := e.GetCurrentState(clusterName)
+	if err != nil {
+		return err
+	}
+
+	resourceProviders, unusedProviders, calcErrs, err := e.CalculateOptimizedState(clusterName, resourceConsumers, resourceProviders)
+	if err != nil {
+		return err
+	}
+
+	desiredScale := len(resourceProviders) - len(unusedProviders)
+	if err := e.ScaleToState(clusterName, desiredScale, unusedProviders); err != nil {
+		return err
+	}
+
+	if len(calcErrs) > 0 {
+		return errors.MultiError(calcErrs)
+	}
+
+	return nil
+}
+
+func (e *EnvironmentScaler) GetCurrentState(clusterName string) ([]scaler.ResourceConsumer, []*scaler.ResourceProvider, error) {
+	resourceProviders, err := e.getResourceProviders(clusterName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	humanReadableProviders := ""
+	for _, r := range resourceProviders {
+		humanReadableProviders += fmt.Sprintf("\n    %#v", r)
```

Comment:
this reads a bit odd to me, what is this spacing for?

Reply:
Just my own subjective readability; zpatrick has a suggestion to make a proper string representation for providers. ✅